src/main/java/org/variantsync/diffdetective/analysis/Analysis.java

```diff
@@ -0,0 +1,351 @@
+
+            // For each commit
+            for (final RevCommit finalCommit : commits) {
+                commit = finalCommit;
+
+                ListIterator<Hooks> commitHook = hooks.listIterator();
+                try {
+                    if (!runFilterHook(commitHook, Hooks::beginCommit)) {
+                        continue;
+                    }
+
+                    processCommit();
+                } catch (Exception e) {
+                    Logger.error(e, "An unexpected error occurred at {} in {}", commit.getId().getName(), repository.getRepositoryName());
+                    throw e;
+                } finally {
+                    runReverseHook(commitHook, Hooks::endCommit);
+                }
+            }
+        } finally {
+            runReverseHook(batchHook, Hooks::endBatch);
+        }
+    }
+
+    protected void processCommit() throws Exception {
+        // parse the commit
+        final CommitDiffResult commitDiffResult = differ.createCommitDiff(commit);
+
+        // report any errors that occurred and exit in case no DiffTree could be parsed.
+        getResult().reportDiffErrors(commitDiffResult.errors());
+        if (commitDiffResult.diff().isEmpty()) {
+            Logger.debug("found commit that failed entirely because:\n{}", commitDiffResult.errors());
+            runHook(hooks.listIterator(), Hooks::onFailedCommit);
+            return;
+        }
+
+        // extract the produced commit diff and inform the strategy
+        commitDiff = commitDiffResult.diff().get();
+        if (!runFilterHook(hooks.listIterator(), Hooks::onParsedCommit)) {
+            return;
+        }
+
+        // inspect every patch
+        for (final PatchDiff finalPatch : commitDiff.getPatchDiffs()) {
+            patch = finalPatch;
+
+            ListIterator<Hooks> patchHook = hooks.listIterator();
+            try {
+                if (!runFilterHook(patchHook, Hooks::beginPatch)) {
+                    continue;
+                }
+
+                processPatch();
+            } finally {
+                runReverseHook(patchHook, Hooks::endPatch);
+            }
+        }
+    }
+
+    protected void processPatch() throws Exception {
+        if (patch.isValid()) {
+            // generate TreeDiff
+            diffTree = patch.getDiffTree();
+            diffTree.assertConsistency();
+
+            runFilterHook(hooks.listIterator(), Hooks::analyzeDiffTree);
+        }
+    }
+
+    protected <Hook> void runHook(ListIterator<Hook> hook, FailableBiConsumer<Hook, Analysis, Exception> callHook) throws Exception {
+        while (hook.hasNext()) {
+            callHook.accept(hook.next(), this);
+        }
+    }
+
+    protected <Hook> boolean runFilterHook(ListIterator<Hook> hook, FailableBiFunction<Hook, Analysis, Boolean, Exception> callHook) throws Exception {
+        while (hook.hasNext()) {
+            if (!callHook.apply(hook.next(), this)) {
+                return false;
+            }
```

Comment:
FilterHooks should be side-effect free but they aren't. For example, `LineGraphExportAnalysis::onParsedCommit` has a side-effect. When exitiving here as soon as some filter says "no", we would omit running the remaining side-effects = bug! So I would argue to just run all hooks and make a big AND of their return values.

Reply:
No. Filter hooks are not required to be side-effect free. I will explain this in the upcoming documentation.

Reply:
As discussed in our last meeting, the current behaviour is fine but should be documented.